LambdaEngine/Source/Game/ECS/Systems/Rendering/AnimationSystem.cpp

```diff
@@ -125,30 +157,41 @@ namespace LambdaEngine
 			transform			= glm::scale(transform, scale);
 
 			// Increase boneID
-			animation.BoneMatrices[boneID] = transform;
+			animation.Pose.LocalTransforms[boneID] = transform;
 		}
-
-		// Apply parent transform
-		TArray<glm::mat4> tempBones = animation.BoneMatrices;
-		for (uint32 i = 0; i < pSkeleton->Bones.GetSize(); i++)
+		
+		// Calculate global transforms
+		for (uint32 i = 0; i < pSkeleton->Joints.GetSize(); i++)
 		{
-			Bone& bone = pSkeleton->Bones[i];
-			tempBones[i] = pSkeleton->GlobalTransform * ApplyParent(bone, *pSkeleton, animation.BoneMatrices) * bone.OffsetTransform;
+			Joint& joint = pSkeleton->Joints[i];
+			animation.Pose.GlobalTransforms[i] = pSkeleton->InverseGlobalTransform * ApplyParent(joint, *pSkeleton, animation.Pose.LocalTransforms) * joint.InvBindTransform;
 		}
-
-		animation.BoneMatrices.Swap(tempBones);
 	}
 
-	glm::mat4 AnimationSystem::ApplyParent(Bone& bone, Skeleton& skeleton, TArray<glm::mat4>& matrices)
+	glm::mat4 AnimationSystem::ApplyParent(Joint& bone, Skeleton& skeleton, TArray<glm::mat4>& matrices)
 	{
 		int32 parentID	= bone.ParentBoneIndex;
-		int32 myID		= skeleton.BoneMap[bone.Name];
-		if (parentID == -1)
+		int32 myID		= skeleton.JointMap[bone.Name];
+		if (parentID == INVALID_JOINT_ID)
 		{
 			return matrices[myID];
 		}
 
-		return ApplyParent(skeleton.Bones[parentID], skeleton, matrices) * matrices[myID];
+		return ApplyParent(skeleton.Joints[parentID], skeleton, matrices) * matrices[myID];
+	}
+
+	void AnimationSystem::OnEntityAdded(Entity entity)
+	{
+		ECSCore* pECSCore = ECSCore::GetInstance();
+		AnimationComponent& animationComp = pECSCore->GetComponent<AnimationComponent>(entity);
+		if (m_HasInitClock)
+		{
+			animationComp.StartTime = GetTotalTimeInSeconds();
+		}
```

Comment:
```suggestion
		if (m_HasInitClock)
		{
		        ECSCore* pECSCore = ECSCore::GetInstance();
		        AnimationComponent& animationComp = pECSCore->GetComponent<AnimationComponent>(entity);
			animationComp.StartTime = GetTotalTimeInSeconds();
		}
```

Reply:
Does not work, "pAnimationComponents" is used further down

Reply:
Looks to me like this is all there is to this function:
https://i.gyazo.com/5e1f3bcfdd6106bffae1f1fc76bb4070.png

Reply:
I looked at wrong function, opps 👀